core/src/main/java/org/apache/iceberg/rest/RESTTableOperations.java

```diff
@@ -169,16 +181,21 @@ public FileIO io() {
     return io;
   }
 
-  private TableMetadata updateCurrentMetadata(LoadTableResponse response) {
-    // LoadTableResponse is used to deserialize the response, but config is not allowed by the REST
-    // spec so it can be
-    // safely ignored. there is no requirement to update config on refresh or commit.
-    if (current == null
-        || !Objects.equals(current.metadataFileLocation(), response.metadataLocation())) {
-      this.current = response.tableMetadata();
+  private void updateCurrentMetadata(LoadTableResponse response) {
+    this.current = response.tableMetadata();
+  }
+
+  private void updateIo(LoadTableResponse response) {
```

Comment:
This is a little difficult because we the Catalog allows for a custom bifunctor to create a new FileIO and that could be doing all sorts of things.

From our offline chat: @eric-maynard suggested 2 options

1.  We use the RestSessionCatalog newFileIO here - This adds a dependency on the Catalog
2. We add a refresh api to FileIO - This requires us adding clone/refresh logic to each IO that needs this behavior

Reply:
I think I'm currently leaning towards FIleIO having a refresh API, then providers who allow for a reconfiguration on the fly can do so or can replace themselves with a new implementation. This also would protect us if other instances are using this same FileIO instance

Reply:
@rdblue Do you have a POV on which of these two approached we should take? I think the PR will need to be pretty much rewritten depending on this decision.

Reply:
Considering the approach of TableOperations class taking the responsibility of refreshing the FileIO credentials, we might hit an issue in Spark engine where FileIO instances are distributed in driver and executor.

From Spark engine POV, 
- During the serialization of iceberg Table instance it uses a wrapper SerializableTableWithSize class. 
- This will skip TableOperations as part of serialization but will include FileIO instance derived from TableOperations.
- Because of this the Table instance in spark executor doesn't have TableOperations but has the FileIO which which driver initially has.
- With the above approach TableOperations will refresh FileIO in spark driver but the spark executor will still have an old FileIO.

Ref 
- Spark plan input partitions: [link](https://github.com/apache/iceberg/blob/main/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkBatch.java#L77)
- Serializable Table: [Link](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/SerializableTable.java#L73-L87)

Reply:
I think that we should go with the first option, which is essentially to pass a function in that can produce a FileIO. That function should reuse previous instances if they do not need to change and register the FileIO with the tracker. I don't think `RESTTableOperations` should be responsible for it.

Please also coordinate with @nastra, who is rebuilding how credentials are passed in the protocol. That should make detecting when to replace the `FileIO` easier.